runtime/src/comm/ugni/comm-ugni.c

```diff
@@ -315,28 +331,52 @@ chpl_comm_pstats_t chpl_comm_pstats;
 
 #define _PSV_VAR(cnt)  chpl_comm_pstats.cnt
 
-#define PERFSTATS_ADD(cnt, val)  (void) _PSV_ADD_FUNC(&_PSV_VAR(cnt), val)
-#define PERFSTATS_INC(cnt)       PERFSTATS_ADD(cnt, 1)
+#define PERFSTATS_LD(cnt)         _PSV_LD_FUNC(&_PSV_VAR(cnt))
+#define PERFSTATS_ST(cnt, val)    _PSV_ST_FUNC(&_PSV_VAR(cnt), val)
+#define PERFSTATS_STZ(cnt)        PERFSTATS_ST(cnt, 0)
+#define PERFSTATS_ADD(cnt, val)   (void) _PSV_ADD_FUNC(&_PSV_VAR(cnt), val)
+#define PERFSTATS_INC(cnt)        PERFSTATS_ADD(cnt, 1)
 
 #include <time.h>
 
 static struct timespec perfstats_timeBase;
 
-static inline uint64_t perfstats_timer_get(void)
+static _PSV_C_TYPE timer_overhead;
+
+static inline _PSV_C_TYPE perfstats_timer_get(void)
 {
   struct timespec _ts;
   (void) clock_gettime(CLOCK_MONOTONIC, &_ts);
-  return (uint64_t) (_ts.tv_sec - perfstats_timeBase.tv_sec) * 1000000000UL
-         + (uint64_t) _ts.tv_nsec;
+  return (_PSV_C_TYPE) (_ts.tv_sec - perfstats_timeBase.tv_sec) * 1000000000UL
+         + (_PSV_C_TYPE) _ts.tv_nsec;
 }
 
-#define PERFSTATS_TIMER_START(ts) uint64_t ts = perfstats_timer_get()
-
-#define PERFSTATS_TIMER_ELAPSED(ts) ((_PSV_C_TYPE) (perfstats_timer_get() - ts))
+#define PERFSTATS_TGET(ts)      ts = perfstats_timer_get()
+#define PERFSTATS_TSTAMP(ts)    _PSV_C_TYPE ts = perfstats_timer_get()
+#define PERFSTATS_TDIFF(te, ts) ((te) - (ts) - timer_overhead)
+#define PERFSTATS_TELAPSED(ts)  PERFSTATS_TDIFF(perfstats_timer_get(), ts)
 
 static void perfstats_init(void)
 {
   (void) clock_gettime(CLOCK_MONOTONIC, &perfstats_timeBase);
+
+  //
+  // Measure the timer overhead.  We only use timer_nsecs so that we
```

Comment:
If you're worried about timer overhead you could consider using `CLOCK_MONOTONIC_COARSE` instead (requires 2.6.32, but that shouldn't be an issue for ugni)

Reply:
Although it's less accurate, so maybe that's not what you want anyways

Reply:
Yeah, these are fairly fine-grained measurements in some cases.